have_sys_socket.c: support for MacOS added

vasiliyk opened the pull request:
Added support for MacOS:

1. CMakeList.txt: check if system is MacOS or Unix/Linux and set "SUPPORT_DISALLOW_SIGNAL_DURING_SENDING"
2. config.h.in: set SUPPORT_DISALLOW_SIGNAL_DURING_SENDING to be accessible for C preprocessor.
3. socket.h: if SUPPORT_DISALLOW_SIGNAL_DURING_SENDING is defined set "config_disallow_signal_during_sending_flag" flag to "SO_NOSIGPIPE" or "MSG_NOSIGNAL"
4. updated "sys_socket_sendto_tcp_target" and "sys_socket_sendto_udp_target" procedures of have_sys_socket.c to use flag config_disallow_signal_during_sending_flag

vasiliyk committed:
have_sys_socket.c: support for MacOS added

goatshriek commented on CMakeLists.txt:
```diff
@@ -294,6 +294,14 @@ endif()
 # configuration-specific source files
+
+# we don't check for win32 as on win32 socket operations are handled directly through return values
+if(APPLE) # MacOS, iOS
```
Rather than basing these checks on the detected configuration platform, please perform the check using `check_symbol_exists` with `SO_NOSIGPIPE` and `MSG_NOSIGNAL`, and then define wrapper based on the resulting `HAVE_SO_NOSIGPIPE` and `HAVE_MSG_NOSIGNAL` `#define` s. This will allow platforms that supply one or the other of these to build correctly, without relying on CMake to bin them as `APPLE` or `UNIX`.

goatshriek commented on include/private/config.h.in:
```diff
@@ -46,6 +46,7 @@
 
 /* function support checks */
 #cmakedefine SUPPORT_ABSTRACT_SOCKET_NAMES 1
+#cmakedefine SUPPORT_DISALLOW_SIGNAL_DURING_SENDING 1
```
This should change to two `HAVE_` `#cmakedefine`s.

goatshriek commented on include/private/config/wrapper/socket.h:
```diff
@@ -41,4 +41,14 @@
 #    define config_get_local_socket_name no_abstract_socket_names_get_local_socket_name
 #  endif
 
+#  ifdef SUPPORT_DISALLOW_SIGNAL_DURING_SENDING
+#    if defined(__APPLE__)
+#          define config_disallow_signal_during_sending_flag SO_NOSIGPIPE
+#    elif defined(__linux__)
+#          define config_disallow_signal_during_sending_flag MSG_NOSIGNAL
+#    elif defined(unix) || defined(__unix__)
+#          define config_disallow_signal_during_sending_flag MSG_NOSIGNAL
+#    endif
+#  endif
+
```
Rework this to use the `HAVE_` `#define`s from the private config.

goatshriek reviewed (changes requested):
This is almost there, just one change outlined below to use CMake's `check_symbol_exists` function to instead determine what symbols are available for use, so that the platform detection doesn't become as much of a dependency for this to work as intended.

vasiliyk committed:
CmakeList.txt: check_symbol_exists used instead of APPLE check is used
HAVE_DISALLOW_SIGNAL_DURING_SENDING definition is used instead of SUPPORT_DISALLOW_SIGNAL_DURING_SENDING

vasiliyk commented:
Joel, thank you for review. The code is updated based on your comments.

goatshriek commented on include/private/config.h.in:
```diff
@@ -46,6 +46,7 @@
 
 /* function support checks */
 #cmakedefine SUPPORT_ABSTRACT_SOCKET_NAMES 1
+#cmakedefine HAVE_DISALLOW_SIGNAL_DURING_SENDING 1
```
You'll need to add the other two `HAVE_` symbols to this file. I don't think you'll need this one anymore - you can use the `HAVE_` ones in the wrapper header instead.

goatshriek commented on include/private/config/wrapper/socket.h:
```diff
@@ -41,4 +41,14 @@
 #  endif
 
+#  ifdef HAVE_DISALLOW_SIGNAL_DURING_SENDING
+#    if defined(__APPLE__)
```
Change this condition to use `HAVE_SO_NOSIGPIPE` and the next to `HAVE_MSG_NOSIGGNAL`.

vasiliyk committed:
removed unnecessary HAVE_DISALLOW_SIGNAL_DURING_SENDING

vasiliyk commented:
Joel, the code was updated based on the last recommendations. Thank you!

goatshriek commented:
Thanks for sticking with this one! There is just one last change needed - an include `sys/socket.h` is needed for the wrapped symbols, which is why the static analyzer has failed. It would probably be _best_ to move the wrapper `#define`s to `include/private/config/have_sys_socket.h` to ensure this header is available for include, but I can take care of this later if you'd prefer to be done. At a minimum, adding this include should resolve the static analysis errors.

vasiliyk committed:
sys/socket.h include added to private/config/have_sys_socket.h

vasiliyk commented:
Joel, I added sys/socket.h include to include/private/config/have_sys_socket.h as you suggested.

goatshriek commented:
Sorry, I should have also mentioned that you'll need to move the code from `...wrapper/socket.h` to `have_sys_socket.h` as well.

vasiliyk committed:
"config_disallow_signal_during_sending_flag" definition moved to have_sys_socket.h

vasiliyk commented:
No problem, I have moved the definition of ``config_disallow_signal_during_sending_flag`` from ``...wrapper/socket.h`` to ``have_sys_socket.h`` as well.

goatshriek commented on src/config/have_sys_socket.c:
```diff
@@ -27,6 +27,7 @@
 #include "private/config/wrapper/locale.h"
 #include "private/config/wrapper/int_connect.h"
 #include "private/config/wrapper/thread_safety.h"
+#include "private/config/wrapper/socket.h"
```
This include is no longer needed with the last change.

goatshriek reviewed (changes requested):
Just one more failing check now - there is an include that is no longer necessary with the last refactoring. I believe once you take out this line it should be good to go.

Thanks again for your patience with this!

vasiliyk committed:
unnecessary include removed

vasiliyk commented:
Joel, I have removed unnecessary include.

goatshriek commented:
Thanks very much for not only raising this issue but putting together a fix and sticking with it through the review process! Portability is a primary goal of this project and it's always great to see progress made towards that.
